hbase-common/src/main/java/org/apache/hadoop/hbase/CellComparatorImpl.java

```diff
@@ -82,6 +86,337 @@ public int compare(final Cell a, final Cell b, boolean ignoreSequenceid) {
+      // given row, because there is no "lexicographically last column" (it would be infinitely
+      // long).
+      // The "maximum" key type does not need this behavior. Copied from KeyValue. This is bad in
+      // that
+      // we can't do memcmp w/ special rules like this.
+      // TODO: Is there a test for this behavior?
+      int leftFamilyLengthPosition = left.getFamilyLengthPosition(leftRowLength);
+      int leftFamilyLength = left.getFamilyLength(leftFamilyLengthPosition);
+      int leftKeyLength = left.getKeyLength();
+      int leftQualifierLength =
+          left.getQualifierLength(leftKeyLength, leftRowLength, leftFamilyLength);
+
+      // No need of left row length below here.
+
+      byte leftType = left.getTypeByte(leftKeyLength);
+      if (leftFamilyLength + leftQualifierLength == 0
+          && leftType == KeyValue.Type.Minimum.getCode()) {
+        // left is "bigger", i.e. it appears later in the sorted order
+        return 1;
+      }
+
+      int rightFamilyLengthPosition = right.getFamilyLengthPosition(rightRowLength);
+      int rightFamilyLength = right.getFamilyLength(rightFamilyLengthPosition);
+      int rightKeyLength = right.getKeyLength();
+      int rightQualifierLength =
+          right.getQualifierLength(rightKeyLength, rightRowLength, rightFamilyLength);
+
+      // No need of right row length below here.
+
+      byte rightType = right.getTypeByte(rightKeyLength);
+      if (rightFamilyLength + rightQualifierLength == 0
+          && rightType == KeyValue.Type.Minimum.getCode()) {
+        return -1;
+      }
+
+      // Compare families.
+      int leftFamilyPosition = left.getFamilyOffset(leftFamilyLengthPosition);
+      int rightFamilyPosition = right.getFamilyOffset(rightFamilyLengthPosition);
+      diff = Bytes.compareTo(left.getFamilyArray(), leftFamilyPosition, leftFamilyLength,
+        right.getFamilyArray(), rightFamilyPosition, rightFamilyLength);
+      if (diff != 0) {
+        return diff;
+      }
+
+      // Compare qualifiers
+      diff = Bytes.compareTo(left.getQualifierArray(),
+        left.getQualifierOffset(leftFamilyPosition, leftFamilyLength), leftQualifierLength,
+        right.getQualifierArray(), right.getQualifierOffset(rightFamilyPosition, rightFamilyLength),
+        rightQualifierLength);
+      if (diff != 0) {
+        return diff;
+      }
+
+      // Timestamps.
+      // Swap order we pass into compare so we get DESCENDING order.
+      diff = Long.compare(right.getTimestamp(rightKeyLength), left.getTimestamp(leftKeyLength));
+      if (diff != 0) {
+        return diff;
+      }
+
+      // Compare types. Let the delete types sort ahead of puts; i.e. types
+      // of higher numbers sort before those of lesser numbers. Maximum (255)
+      // appears ahead of everything, and minimum (0) appears after
+      // everything.
+      return (0xff & rightType) - (0xff & leftType);
+    } else if (l instanceof KeyValue && r instanceof ByteBufferKeyValue) {
```

Comment:
For     } else if (l instanceof ByteBufferKeyValue && r instanceof KeyValue) {
and } else if (l instanceof KeyValue && r instanceof ByteBufferKeyValue) { cases, essentially, there are same.
Can we have a method for BBKVCompare() and swap parameters for these two cases?